File: Sources/Runtime/DevClient/HotReloadClient.swift
#if MELODY_DEV
import Foundation
import Observation
import Core

@MainActor
@Observable
/// WebSocket client that receives YAML updates from the dev server and triggers live reloads.
public final class HotReloadClient {
    public private(set) var latestApp: AppDefinition?
    public private(set) var isConnected = false
    public private(set) var reloadCount = 0

    private var webSocketTask: URLSessionWebSocketTask?
    private var session: URLSession?
    private let parser = AppParser()
    private var shouldReconnect = true

    public init() {}

    /// Connect to the dev server
    public func connect(host: String = "localhost", port: Int = 8375) {
        shouldReconnect = true
        let url = URL(string: "ws://\(host):\(port)")!
        session = URLSession(configuration: .default)
        startConnection(url: url)
    }

    /// Disconnect from the dev server
    public func disconnect() {
        shouldReconnect = false
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        isConnected = false
    }

    // MARK: - Private

    private func startConnection(url: URL) {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        let task = session!.webSocketTask(with: url)
        self.webSocketTask = task
        task.resume()
        isConnected = true
        DevLogger.shared.log("Connected to \(url)", source: "hotreload")
        receiveMessage(url: url)
    }

    private func receiveMessage(url: URL) {
        webSocketTask?.receive { [weak self] result in
            Task { @MainActor [weak self] in
                guard let self else { return }

                switch result {
                case .success(let message):
                    switch message {
                    case .string(let yaml):
                        self.handleYamlUpdate(yaml)
                    case .data(let data):
                        if let yaml = String(data: data, encoding: .utf8) {
                            self.handleYamlUpdate(yaml)
                        }
                    @unknown default:
                        break
                    }
                    self.receiveMessage(url: url)

                case .failure(let error):
                    self.isConnected = false
                    DevLogger.shared.log("Connection failed: \(error.localizedDescription)", source: "hotreload")
                    if self.shouldReconnect {
                        try? await Task.sleep(nanoseconds: 2_000_000_000)
                        guard self.shouldReconnect else { return }
                        self.startConnection(url: url)
                    }
                }
            }
        }
    }

    private func handleYamlUpdate(_ rawPayload: String) {
        var yaml = rawPayload
        var reason: String?

        // Parse optional reload reason header: "---melody-reload: <reason>\n"
        let prefix = "---melody-reload: "
        if yaml.hasPrefix(prefix), let newline = yaml.firstIndex(of: "\n") {
            reason = String(yaml[yaml.index(yaml.startIndex, offsetBy: prefix.count)..<newline])
            yaml = String(yaml[yaml.index(after: newline)...])
        }

        do {
            let app = try parser.parse(yaml)
            Task { @MainActor in
                self.latestApp = app
                self.reloadCount += 1
                let detail = reason.map { " — \($0)" } ?? ""
                DevLogger.shared.log("Reload #\(self.reloadCount)\(detail)", source: "hotreload")
            }
        } catch {
            DevLogger.shared.log("Parse error: \(error.localizedDescription)", source: "hotreload")
        }
    }
}
#endif

File: Android/melody-runtime/src/main/java/com/melody/runtime/devclient/HotReloadClient.kt
package com.melody.runtime.devclient

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.melody.core.parser.AppParser
import com.melody.core.schema.AppDefinition
import okhttp3.*
import okhttp3.internal.toCanonicalHost

/**
 * Connects to the Melody dev server and receives YAML updates over WebSocket.
 * Port of iOS HotReloadClient.swift.
 *
 * Usage: create in a composable, call connect(), observe latestApp and reloadCount.
 */
class HotReloadClient {
    var latestApp: AppDefinition? by mutableStateOf(null)
        private set
    var isConnected: Boolean by mutableStateOf(false)
        private set
    var reloadCount: Int by mutableIntStateOf(0)
        private set

    private val parser = AppParser()
    private val client = OkHttpClient()
    private var webSocket: WebSocket? = null
    private var shouldReconnect = true
    private var connectUrl: String? = null

    fun connect(host: String = "10.0.2.2", port: Int = 8375) {
        shouldReconnect = true
        connectUrl = "http://$host:$port"
        startConnection()
    }

    fun disconnect() {
        shouldReconnect = false
        webSocket?.close(1000, "Client disconnected")
        webSocket = null
        isConnected = false
    }

    private fun startConnection() {
        val url = connectUrl ?: return

        try {
            java.net.URL(url).host?.takeIf { it.isNotEmpty() }
        } catch (_: Exception) {
            DevLogger.log("Invalid URL: $url", "hotreload")
            Log.e("Melody", "[HotReload] Invalid URL: $url")
            null
        } ?: return

        val request = Request.Builder().url(url).build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                isConnected = true
                DevLogger.log("Connected to $url", "hotreload")
                Log.d("Melody", "[HotReload] Connected to $url")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleYamlUpdate(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                isConnected = false
                DevLogger.log("Connection failed: ${t.message}", "hotreload")
                Log.w("Melody", "[HotReload] Connection failed: ${t.message}")

                if (shouldReconnect) {
                    android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
                        if (shouldReconnect) startConnection()
                    }, 2000)
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                isConnected = false
                if (shouldReconnect) {
                    android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
                        if (shouldReconnect) startConnection()
                    }, 2000)
                }
            }
        })
    }

    private fun handleYamlUpdate(yaml: String) {
        try {
            val app = parser.parse(yaml)
            latestApp = app
            reloadCount++
            DevLogger.log("Reload #$reloadCount", "hotreload")
            Log.d("Melody", "[HotReload] Reload #$reloadCount")
        } catch (e: Exception) {
            DevLogger.log("Parse error: ${e.message}", "hotreload")
            Log.e("Melody", "[HotReload] Parse error: ${e.message}")
        }
    }
}
